### PR TITLE
Refactor: claim service to TS + refactoring

### PR DESCRIPTION
**Pull request by adrien2p**
**Goal**
- Move from JS to TS

**Improvements**
- Improve typings when required
- Reduce the query amount when possible such as avoid multiple save when only one could be done instead
- Use `Promise.all` when makes sense instead of `for..of` loop to take advantage of I/O operations

**Fixes**
- Add missing await from return such as the following example (but thats not all) `return this.atomicPhase_` becomes `return await this.atomicPhase_`

### CODE REVIEW

File: packages/medusa/src/services/claim.ts [hunk 1]
```diff
@@ -91,55 +163,57 @@ class ClaimService extends BaseService {
...
+        await Promise.all(
+          shipping_methods.map(shippingMethod => {
+            if (shippingMethod.id) {
+              return this.shippingOptionService_
+                .withTransaction(transactionManager)
+                .updateShippingMethod(shippingMethod.id, {
+                  claim_order_id: claim.id,
+                })
+            } else if (shippingMethod.option_id) {
+              // TODO: data does not exists on shippingMethod, should have a look
```
Comment (adrien2p) [inline packages/medusa/src/services/claim.ts:194]: @olivermrbl data does not exists there

File: packages/medusa/src/services/claim.ts [hunk 2]
```diff
@@ -274,89 +352,93 @@ class ClaimService extends BaseService {
...
               })
-          } else {
+          } else if (method.option_id) {
+            // TODO: data does not exists on shippingMethod, should have a look
             await this.shippingOptionService_
-              .withTransaction(manager)
-              .createShippingMethod(method.option_id, method.data, {
-                claim_order_id: result.id,
+              .withTransaction(transactionManager)
+              .createShippingMethod(method.option_id, (method as any).data, {
```
Comment (adrien2p) [inline packages/medusa/src/services/claim.ts:383]: @olivermrbl data does not exists there

File: packages/medusa/src/services/claim.ts [hunk 3]
```diff
@@ -274,89 +352,93 @@ class ClaimService extends BaseService {
...
-          claim_order_id: result.id,
-          items: claim_items.map((ci) => ({
-            item_id: ci.item_id,
-            quantity: ci.quantity,
-            metadata: ci.metadata,
+          claim_order_id: claimOrder.id,
+          items: claim_items.map((claimItem) => ({
+            item_id: claimItem.item_id,
+            quantity: claimItem.quantity,
+            metadata: (claimItem as any).metadata,
```
Comment (adrien2p) [inline packages/medusa/src/services/claim.ts:406]: @olivermrbl data does not exists there, it is an `Item`

File: packages/medusa/src/services/claim.ts [hunk 4]
```diff
@@ -1,8 +1,51 @@
-import { MedusaError } from "medusa-core-utils"
+import ClaimItemService from "./claim-item"
```
Comment (srindom) [inline packages/medusa/src/services/claim.ts:1]: Please double-check if your linter is configured correctly; looks like some of the formatting is a bit off :)

Comment (adrien2p) [reply packages/medusa/src/services/claim.ts:1]: I just re lint and add the lint on save from the closest location ;)

Comment (adrien2p) [reply packages/medusa/src/services/claim.ts:1]: could you have a look at the comment above please :)

**Comment by adrien2p** [comment]
@srindom is it fine for you?

**Comment by adrien2p** [comment]
@olivermrbl this one just need to re run the pipeline please

**Review by srindom — APPROVED** [review]
LGTM!

**Comment by adrien2p** [comment]
> LGTM!

@srindom could you look at the comments I did put :)

**Comment by olivermrbl** [comment]
@adrien2p Why is this a bug?

**Comment by adrien2p** [comment]
> @adrien2p Why is this a bug?

I ve added bug because there is some missing atomicPhase which means that depending the usage some query could end up out of a transaction

File: packages/medusa/src/services/claim.ts [hunk 5]
```diff
@@ -0,0 +1,872 @@
...
+    claimId: string,
+    config: FindConfig<ClaimOrder> = {}
+  ): Promise<ClaimOrder> {
+    return await this.atomicPhase_(
+      async (transactionManager: EntityManager) => {
+        const claimRepo = transactionManager.getCustomRepository(
+          this.claimRepository_
+        )
+
+        const validatedId = validateId(claimId)
```
Comment (olivermrbl) [inline packages/medusa/src/services/claim.ts:857]: suggestion(non-blocking): `validateId` does not have any effect when used without a config, so I would remove the use of it here
```suggestion
```

**Review by olivermrbl — APPROVED** [review]
LGTM! 💪

File: packages/medusa/src/services/claim.ts [hunk 6]
```diff
@@ -0,0 +1,872 @@
...
+   * @return {Object} created claim
+   */
+  async create(
+    data: AdminPostOrdersOrderClaimsReq & {
+      order: Order
+      type?: string
+      claim_order_id?: string
+      shipping_address_id?: string
+    }
+  ): Promise<ClaimOrder> {
```
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/services/claim.ts:230]: Yes true 👌

**Pull request merged**